greeter/README.md

```diff
@@ -1,21 +1,510 @@
+    ) {}
+    ```
+
+    And the following import statements:
+
+    ```ts
+    import {inject} from '@loopback/context';
+    import {CachingService} from '../services/caching.service';
+    import {CACHING_SERVICE} from '../keys';
+    ```
+
+3. For the `start()` and `stop()` function, we want to start and stop the caching service.
+
+    ```ts
+    /**
+     * This method will be invoked when the application starts
+     */
+    async start(): Promise<void> {
+        // Add your logic for start
+        await this.cachingService.start();
+    }
+
+    /**
+     * This method will be invoked when the application stops
+     */
+    async stop(): Promise<void> {
+        // Add your logic for stop
+        await this.cachingService.stop();
+    }
+    ```
+
+
+### Step 3: Create global interceptor for caching
+
+1. Run the interceptor generator `lb4 interceptor` command.
+
+    ```sh
+    $ lb4 interceptor
+    ? Interceptor name: caching
+    ? Is it a global interceptor? Yes
+
+    Global interceptors are sorted by the order of an array of group names bound to ContextBindings.GLOBAL_INTERCEPTOR_ORDERED_GROUPS. See https://loopback.io/doc/en/lb4/Interceptors.html#order-of-invocation-for-interceptors.
+
+    ? Group name for the global interceptor: ('') 
+    create src/interceptors/caching.interceptor.ts
+    update src/interceptors/index.ts
+
+    Interceptor Caching was created in src/interceptors/
+    ```
+
+2. In `src/interceptors/caching.interceptors.ts`, modify the constructor:
+
+    ```ts
+    constructor(
+        @inject(CACHING_SERVICE) private cachingService: CachingService,
+    ) {}
+    ```
+
+    Add and modify the following import statements:
+
+    ```ts
+    import {
+        /* inject, */
+        globalInterceptor,
+        Interceptor,
+        InvocationContext,
+        InvocationResult,
+        Provider,
+        inject,
+        ValueOrPromise,
+    } from '@loopback/context';
+    import {CachingService} from '../services/caching.service';
+    import {CACHING_SERVICE} from '../keys';
+    import {RestBindings} from '@loopback/rest';
+    ```
+
+3. Add the pre-invocation and post-invocation logics in the `intercept` function.
+
+    ```ts
+    // Add pre-invocation logic here
+    const httpReq = await invocationCtx.get(RestBindings.Http.REQUEST, {
+        optional: true,
+    });
+    /* istanbul ignore if */
+    if (!httpReq) {
+        // Not http request
+        return next();
+    }
+    const key = httpReq.path;
+    const lang = httpReq.acceptsLanguages(['en', 'zh']) || 'en';
+    const cachingKey = `${lang}:${key}`;
+    const cachedResult = await this.cachingService.get(cachingKey);
+    if (cachedResult) {
+        console.error('Cache found for %s %j', cachingKey, cachedResult);
+        return cachedResult;
+    }
+
+    const result = await next();
+
+    // Add post-invocation logic here
+    await this.cachingService.set(cachingKey, result);
+    return result;
+    ```
+
```

Comment:
Missing one step: bind interceptor and service to app

```ts
 this.add(createBindingFromClass(CachingService, {key: CACHING_SERVICE}));
    this.add(createBindingFromClass(CachingInterceptor));
```

Reply:
in application.ts we would need to add:

```
import {CachingService} from './caching.service';
import {CACHING_SERVICE} from './keys';
import {CachingInterceptor} from './interceptors/caching.interceptor';

```

Reply:
I am getting this in the output:

```
Server is running at http://[::1]:3000
Try http://[::1]:3000/ping
TypeError: Cannot read property 'getTime' of undefined
    at CachingService.isExpired (/Users/dremond/Documents/PRs_LoopBack/workshop/theapp/greeter/src/caching.service.ts:83:42)
    at process._tickCallback (internal/process/next_tick.js:68:7)
TypeError: Cannot read property 'getTime' of undefined
    at CachingService.isExpired (/Users/dremond/Documents/PRs_LoopBack/workshop/theapp/greeter/src/caching.service.ts:83:42)
    at process._tickCallback (internal/process/next_tick.js:68:7)
TypeError: Cannot read property 'getTime' of undefined
    at CachingService.isExpired (/Users/dremond/Documents/PRs_LoopBack/workshop/theapp/greeter/src/caching.service.ts:83:42)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Reply:
![image](https://user-images.githubusercontent.com/6864736/66952260-46491c80-f02a-11e9-905d-7472378812c5.png)

openapi data?

I think we should check for  `msg.timestamp` as well as 'msg'

Reply:
Also let's make caching last a little longer. OR we tell user to press `Execute` many times quickly to notice it is happening, and when it expires and starts again.